### PR TITLE
upgrade, upgrade/v100to101: do not update IM label if it is up to date, add retry logic when doing CRDs update

### PR DESCRIPTION
**Pull request by PhanLe1010**
1. To reduce the number of conflicts when doing CRDs upgrade, we do not update IM's label if it is up to date. This fix alone solves the issue.
2. Adding retry logic when doing CRDs update so that one API conflict does not block Longhorn upgrading. This fix alone also solves the issue.

We combine both of the above fixes to achieve fast (reduce the number of conflicts) and safe (adding retry logic).

longhorn/longhorn#1646

### CODE REVIEW

File: upgrade/v100to101/upgrade.go [hunk 1]
```diff
@@ -92,7 +92,8 @@ func doInstanceManagerUpgrade(namespace string, lhClient *lhclientset.Clientset)
 	}
 
 	for _, im := range imList.Items {
-		if im.Spec.Image != "" {
+		newInstanceManagerLabels := types.GetInstanceManagerLabels(im.Spec.NodeID, im.Spec.Image, im.Spec.Type)
```
Comment (yasker) [inline upgrade/v100to101/upgrade.go:95]: Can we split this part into another commit? It's specific for v1.0.0 to v1.0.1 upgrade.

File: upgrade/upgrade.go [hunk 2]
```diff
@@ -102,16 +104,27 @@ func upgrade(currentNodeID, namespace string, config *restclient.Config, lhClien
...
+				})
+				if err != nil {
 					return
 				}
-				if err = doCRDUpgrade(namespace, lhClient); err != nil {
+
+				err = retryOnConflictCause(func() error {
+					return doCRDUpgrade(namespace, lhClient)
+				})
+				if err != nil {
```
Comment (yasker) [inline upgrade/upgrade.go:124]: What if the error is not caused by the conflict? Then we will return and stuck as the leader for upgrade forever again.

File: upgrade/v100to101/upgrade.go (hunk 1 above)
Comment (PhanLe1010) [reply upgrade/v100to101/upgrade.go:95]: fixed

File: upgrade/upgrade.go (hunk 2 above)
Comment (PhanLe1010) [reply upgrade/upgrade.go:124]: Yes, you are right. And I think that it makes sense because if the error is not caused by the conflict, something else is wrong in the upgrading process (some CRDs with new settings while other with old settings) and should stop Longhorn from working. 

If we want Longhorn managers to keep going even if there is error in one Longhorn manager, we can `defer cancel()`. I tested it yesterday. All manager will keep going even though there are errors. 

How do you think?

Comment (yasker) [reply upgrade/upgrade.go:124]: `defer cancel()` is a good idea. It will not block other managers and the one always fail will end up in the leader and won't cause more harm to the system.

Comment (PhanLe1010) [reply upgrade/upgrade.go:126]: But in case the Longhorn system with only 1 manager, the manger will keep going even when it fails to upgrade. I guess this is the conner case though.

Sure, added the `defer cancel()`.

Comment (yasker) [reply upgrade/upgrade.go:126]: hmm? it should fail (crash) rather than keep going.

Comment (PhanLe1010) [reply upgrade/upgrade.go:126]: No. It keeps going ... because there is no error returned from https://github.com/longhorn/longhorn-manager/blob/5b1bcac398d7a6b7412a56b5fbe0389f6fa9673b/upgrade/upgrade.go#L76

Comment (yasker) [reply upgrade/upgrade.go:126]: Then we should fix that. Fail to pass the upgrade period should crash the process and let it retry.

Comment (PhanLe1010) [reply upgrade/upgrade.go:126]: Yes, I will fix it

**Comment by PhanLe1010** [comment]
close this one because open an other PR for this issue at https://github.com/longhorn/longhorn-manager/pull/637
